Support ingestion of long/float dimensions

jon-wei opened the pull request:
This PR adds long and float implementations of DimensionHandler/DimensionIndexer/DimensionMerger, allowing ingestion of long/float typed dimensions.

This also changes the `EncodedTypeArray` type parameter in the interfaces above to `EncodedKeyComponentType`, removing the restriction that the individual fields of row keys during ingestion must be arrays (to allow for single numeric values in the keys, since long/floats don't support multivalue rows)

gianm commented on processing/src/main/java/io/druid/segment/LongDimensionIndexer.java:
```diff
@@ -0,0 +1,143 @@
+  @Override
+  public Object convertUnsortedEncodedKeyComponentToActualArrayOrList(Long key, boolean asList)
+  {
+    return Lists.newArrayList(key);
```
ImmutableList.of may be cheaper. It just stores the `key` as a field.

jon-wei replied:
Changed to use ImmutableList.of

gianm commented on processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java:
```diff
@@ -0,0 +1,143 @@
+  @Override
+  public Float getMinValue()
+  {
+    return 0.0f;
```
Similar comment to min/max in the Long handler.

jon-wei replied:
Changed to use min/max here as well

gianm commented on processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java:
```diff
@@ -0,0 +1,143 @@
+  @Override
+  public Indexed<Float> getSortedIndexedValues()
+  {
+    return null;
```
Similar comment to getSortedIndexedValues in the Long handler.

jon-wei replied:
Changed this and the bitmap-related method to throw UnsupportedOperationException

gianm commented on processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java:
```diff
@@ -0,0 +1,143 @@
+  @Override
+  public Object convertUnsortedEncodedKeyComponentToActualArrayOrList(Float key, boolean asList)
+  {
+    return Lists.newArrayList(key);
```
Similar comment to convertUnsortedEncodedKeyComponentToActualArrayOrList in the Long handler.

jon-wei replied:
Changed to use ImmutableList.of

jon-wei commented on docs/content/ingestion/index.md:
```diff
@@ -36,7 +36,21 @@ An example dataSchema is shown below:
       "dimensionsSpec" : {
-        "dimensions": ["page","language","user","unpatrolled","newPage","robot","anonymous","namespace","continent","country","region","city"],
+        "dimensions": [
+          "page","language","user","unpatrolled","newPage","robot","anonymous","namespace","continent","country","region","city",
```
Moved these to new lines

jon-wei commented on docs/content/ingestion/index.md:
```diff
@@ -36,7 +36,21 @@ An example dataSchema is shown below:
+        "dimensions": [
+          "page","language","user","unpatrolled","newPage","robot","anonymous","namespace","continent","country","region","city",
+          {
+            "type": "LONG",
```
They can be lowercase, made them lowercase in the docs now

jon-wei commented on docs/content/ingestion/index.md:
```diff
@@ -169,10 +183,38 @@ handle all formatting decisions on their own, without using the ParseSpec.
 | Field | Type | Description | Required |
 |-------|------|-------------|----------|
-| dimensions | JSON String array | The names of the dimensions. If this is an empty array, Druid will treat all columns that are not timestamp or metric columns as dimension columns. | yes |
+| dimensions | JSON Object array | A list of [dimension schema](#dimension-schema) objects or dimension names. Providing a name is equivalent to providing a String-typed dimension schema with the given name. If this is an empty array, Druid will treat all columns that are not timestamp or metric columns as String-typed dimension columns. | yes |
```
Changed to "JSON array"

jon-wei commented on docs/content/ingestion/index.md:
````diff
@@ -169,10 +183,38 @@ handle all formatting decisions on their own, without using the ParseSpec.
+```json
+"dimensionsSpec" : {
+  "dimensions": [
+    "page","language","user","unpatrolled","newPage","robot","anonymous","namespace","continent","country","region","city",
````
Moved dimensions to new lines

jon-wei commented on processing/src/main/java/io/druid/segment/LongDimensionIndexer.java:
```diff
@@ -0,0 +1,143 @@
+  @Override
+  public Indexed<Long> getSortedIndexedValues()
+  {
+    return null;
```
Changed this and the bitmap-related method to throw UnsupportedOperationException

jon-wei commented on processing/src/main/java/io/druid/segment/LongDimensionIndexer.java:
```diff
@@ -0,0 +1,143 @@
+  @Override
+  public Long getMinValue()
+  {
+    return 0L;
```
Changed these to use MIN_VALUE and MAX_VALUE

jon-wei commented:
Re: your comment on makeColumnValueSelector

thanks, I missed that, I added a couple of tests, one in GroupBy and one in TopN that have aggregations on numeric columns ingested as dimensions, these trigger the float->long and long->float cases you mentioned in IncrementalIndexStorageAdapter

I added something to makeLong/FloatColumnSelector to create wrapping selectors that cast the values returned by the base selector from the indexer if necessary

gianm commented:
@jon-wei The wrapping stuff in IISA feels a bit weird -- I think it'd be nicer for the DimensionIndexers to handle that on their own. So, that means replacing makeColumnValueSelector with makeObjectColumnSelector, makeLongColumnSelector, makeFloatColumnSelector, and makeDimensionSelector, giving full control over behavior to the type handler.

The long/float handlers should probably return NullDimensionSelector for makeDimensionSelector since QISA behaves like that too.

The string handler should probably return Zero-selectors for long/float since, again, QISA behaves like that.

gianm commented:
Also, the current code in your patch throws UnsupportedOperationExceptions in some cases, but I think we should avoid those and return null/zero selectors instead.

fjy commented:
👍

leventov commented on processing/src/main/java/io/druid/segment/StringDimensionIndexer.java:
```diff
@@ -539,7 +546,70 @@ public int lookupId(String name)
+      IncrementalIndexStorageAdapter.EntryHolder currEntry, IncrementalIndex.DimensionDesc desc
+  )
+  {
+     return ZeroLongColumnSelector.instance();
```
How valid is it to return zero selector here? Maybe try to convert from String to number?

gianm replied:
I believe QueryableIndexStorageAdapter does the same thing, so at least it's consistent with that. However, it'd be nice (for schema evolution of a column from string -> numeric) for the behavior to change to try to convert the string to a number.

leventov commented on processing/src/main/java/io/druid/segment/StringDimensionIndexer.java:
```diff
@@ -539,7 +546,70 @@ public int lookupId(String name)
+      IncrementalIndexStorageAdapter.EntryHolder currEntry, IncrementalIndex.DimensionDesc desc
+  )
+  {
+    return ZeroFloatColumnSelector.instance();
```
Same

gianm replied:
Same as https://github.com/druid-io/druid/pull/3966#discussion_r125680562.

licl2014 commented:
@jon-wei @gianm  i have a doubt that long/float dimensions will cause high Cardinality  easily when we have a group by query.so have we consider this scenario?

gianm commented:
@licl2014 it could happen, although in that case Druid will do the best it can. If cardinality is not super high it should be fine and go quickly. If cardinality is super high then Druid will spill to disk as needed (assuming disk spilling is enabled).
